### ISSUE TITLE
Mysterious consumer stop in 2.0.3 from draining offset batches on failure

### ISSUE THREAD

**Issue by jyates** [akka/alpakka-kafka#1159]
# Versions used

Scala 2.11.12
Akka: 2.5.[29, 30]
Alpakka-kafka: 2.0.3 and 2.0.3+[2-7].
kafka: 5.4.2-ccs

### Expected Behavior

We are running 2.0.2 in production without issue. 

### Actual Behavior

With 2.0.3 we are seeing the source logging
```
 a.k.i.SingleSourceLogic:92 - [34665] Completing
```

And the consumer stopping. This happens only on a couple of our consumers, but for no obvious reason. No stack traces, nothing untoward in the logs.

 It is bad enough that the consumers stop making tangible progress due to the frequency of rebalances. We see this much more in some consumer groups than others, but it is reproducible in the same consumer, i.e. it seems somehow related to processing patterns.

When turning on DEBUG logging, the issue still occurs and no obvious stack traces or other issues are logged. However, when turning on actor message debugging, ie.
```
akka.actor.debug {
      receive = true
      autoreceive = true
      lifecycle = true
      unhandled = true
}
```
the issue goes away for much longer (10s of minutes rather than within seconds), but still appears. `autorecieve=true` appears to be the main determinant of time until the issue appears.

### Relevant logs

Redacted, DEBUG (akka.kafka, but not actor message handling) logs
https://gist.github.com/jyates/e9b761f14c05a1fff78e08e9262203fa

### Reproducible Test Case

Unfortunately, no obvious test case as its rare even in prod.

**Comment by jyates** [comment]
After walking the commit history, through production (only reliable place to reproduce), it appears the issue came from #1058 . Still have to dig a big more to figure out root cause, but any hints would be helpful if folks get a chance.

**Comment by jyates** [comment]
I'm convinced now that there is a bug in the `CommitCollectorStage`. Commented out and running in prod for much longer than any other time taken to see the issue (often just minutes) has proven to cause no problems, ie. the change
```diff
@@ -28,8 +28,9 @@ object Committer {
    */
   def batchFlow(settings: CommitterSettings): Flow[Committable, CommittableOffsetBatch, NotUsed] = {
     val offsetBatches: Flow[Committable, CommittableOffsetBatch, NotUsed] =
-      Flow
-        .fromGraph(new CommitCollectorStage(settings))
+      Flow[Committable]
+        .groupedWeightedWithin(settings.maxBatch, settings.maxInterval)(_.batchSize)
+        .map(CommittableOffsetBatch.apply)
 
     // See https://github.com/akka/alpakka-kafka/issues/882
     import akka.kafka.CommitDelivery._
```
seems to pinpoint the problem in the `CommitCollectorStage`.

Now, where that problem is, that is another question. Custom stages are hard to get right and harder to debug when things go wrong :-/

Wondering if

**Comment by jyates** [comment]
@ennru, @seglo or @sdudzin maybe you have some ideas?

**Comment by seglo** [comment]
Thanks for the report @jyates.  I can't think of anything off the top of my head, but I'll dig into the logs you provided tomorrow.

**Comment by jyates** [comment]
Ah, should have let it sit a while longer. I think the issue is here
```diff
   override protected def onTimer(timerKey: Any): Unit = timerKey match {
-    case CommitCollectorStage.CommitNow => pushDownStream(Interval)(push)
+    case CommitCollectorStage.CommitNow => pushDownStream(Interval)(emit[CommittableOffsetBatch])
   }
```
When running this path, I'm not seeing any `Completing` from hundreds of instances consuming hundreds of events per second. Moreover, when looking back at a bunch of the logs from times that it failed (and various debugging attempts), I always see something exactly like

```
2020-07-15 18:53:08 DEBUG [KafkaRawToCanonical-akka.actor.default-dispatcher-5] a.k.i.CommitCollectorStageLogic:97 - [20633] pushDownStream triggered by interval, outstanding batch CommittableOffsetBatch(batchSize=20, GroupTopicPartition(group,foo,59) -> 12122501, GroupTopicPartition(group,foo,63) -> 12123060)
2020-07-15 18:53:08 DEBUG [KafkaRawToCanonical-akka.actor.default-dispatcher-5] a.k.i.CommitCollectorStageLogic:97 - [20633] CommitCollectorStage stopped
2020-07-15 18:53:08 INFO  [KafkaRawToCanonical-akka.actor.default-dispatcher-5] a.k.i.SingleSourceLogic:92 - [751bd] Completing
```

I can't say exactly why `emit` is necessary though, as an active batch should only be triggerable by a downstream request being 'passed through' to the upstream for `Committable`s...but then again prod doesn't lie :)

I've had similar issues like this before on custom stages i've written, so this isn't totally crazy, just hard to nail down the exact conditions that lead to the race.

**Comment by seglo** [comment]
I wonder if an exception is being swallowed during a timer triggered commit.  If the timer ticks before the outlet is pulled from upstream, then an `IllegalStateException` will be thrown, which will cause the stage and stream to fail.  I would expect that error to propagate to the stream mat value, where it can be logged by the user, but it might not show up in any logs if that's not done.

As an experiment you could put some exception handling around the timer triggered commit.  If it's indeed a case of pushing to an outlet twice then we can probably just skip the commit if it's not been pulled yet, and try again in the next tick.

**Comment by jyates** [comment]
Good call. Wrapping that call gives me `java.lang.IllegalArgumentException: Cannot push port (FlowOut(1224914437)) twice, or before it being pulled`. 

I think `emit()` solves the problem a bit more elegantly though, than dropping based on pulls. Seems like exactly the kind of case talked about here: https://akka.io/blog/2016/10/21/emit-and-friends

**Comment by seglo** [comment]
I think `emit` makes sense in the case of stream completion where we just want to emit the batch downstream to drain the stream, ahead of completing the stage.  The `CommitCollectorStage` already does this, and AFAICT it's the also the scenario that's being described in the blog post.  

The timer tick's are just part of the `CommitCollectorStages` normal operation.  If there's no demand downstream then I think it would be acceptable to skip an interval and try to commit on the next one.  The consequence of doing this is that the batch could get a little bigger, but that might be fine.

`emit` would also work, but it I'm not sure what the consequences of the extra overhead the [scaladocs warn about](https://doc.akka.io/api/akka/2.6/akka/stream/stage/GraphStageLogic.html#emit[T](out:akka.stream.Outlet[T],elem:T,andThen:()=>Unit):Unit) would be when we emit frequently.  The `CommitCollectorStage` was created to avoid extra allocations, over a simpler operator DSL implementation that was originally proposed, so I just want to make sure we don't impact its efficiency.

**Comment by jyates** [comment]
If we just wait for the next pull/push cycle, we haven't actually followed the guidelines which is to flush every interval, when the downstream is ready. If we just skip the push on this cycle we need to wait for the next record from the upstread to commit. However, that record may never occur. So our batch does not get filled and we don't flush until the next timer cycle occurs. So now we are two intervals behind.

Timer based flushes should be relatively rare in practice. Either you have reasonable intervals between flushes (multiple minutes) or you are already paying a large overhead by flushing with very high frequency anyways. So `emit` _should_ have a relatively low impact.

**Comment by seglo** [comment]
> If we just skip the push on this cycle we need to wait for the next record from the upstread to commit. However, that record may never occur. So our batch does not get filled and we don't flush until the next timer cycle occurs. So now we are two intervals behind.

That's true, but the actual size of a commit batch isn't necessarily what is batched by `CommitCollectorStage`.  The `KafkaConsumerActor` does its own batching (batching of batches!) because we only commit offsets every `poll-interval` to avoid too many in-flight commit requests.  Also, if there's no demand downstream then even if we were to `emit` the batch I don't think that the downstream stage to send that batch to the `KafkaConsumerActor` will pull it until it's ready anyway.

All that said, I'm not necessarily opposed to using `emit` in this case.  It would be the simpler option.  But I want to make sure it's the idiomatic thing to do given that the stream isn't in a state where we know it's shutting down.  I'll ask some people internally.

**Comment by seglo** [comment]
I think using `emit` is ok.  The implementation will only actually do the `emit` behaviour when the port isn't available (as in the case where you observed an error at runtime), so I don't expect the chain of out handlers that emit creates in this scenario to be too long, or happen very regularly.  Plus, as you already mentioned, batches should have a reasonable interval anyway.

The caveat is that we are storing the batch(es) around indefinitely in memory until there is demand from downstream, but I guess we would do that anyway if we wanted to avoid using `emit` and buffer it internally in this stage, or just keep adding to the batch itself.  Due to back-pressure I'm not sure there would ever be a scenario where we would have more than one batch buffered anyway.

**Comment by seglo** [comment]
Fixed in #1160.
